File: impl/src/main/scala/org/gtri/util/iteratee/impl/Iteratee.scala
/*
    Copyright 2012 Georgia Tech Research Institute

    Author: lance.gatlin@gtri.gatech.edu

    This file is part of org.gtri.util.iteratee library.

    org.gtri.util.iteratee library is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    org.gtri.util.iteratee library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with org.gtri.util.iteratee library. If not, see <http://www.gnu.org/licenses/>.

*/

package org.gtri.util.iteratee.impl

import collection.immutable.Iterable
import org.gtri.util.iteratee.api.Issue

/**
 * Created with IntelliJ IDEA.
 * User: Lance
 * Date: 11/11/12
 * Time: 10:27 PM
 * To change this template use File | Settings | File Templates.
 */
object Iteratee {

  type Chunk[A] = Iterable[A]
  type Result[V] = (Option[V], List[Issue])
  type In[A] = (Chunk[A], List[Issue])
  type Iteratee[A,V] = scalaz.IterV[In[A], Result[V]]
  type Input[A] = scalaz.Input[In[A]]

  object EOF {
    def apply[A] : Input[A] = scalaz.IterV.EOF[In[A]]
    def unapply[A](r: Input[A]): Boolean = scalaz.IterV.EOF.unapply(r)
  }

  object Empty {
    def apply[A] : Input[A] = scalaz.IterV.Empty[In[A]]
    def unapply[A](r: Input[A]): Boolean = scalaz.IterV.Empty.unapply(r)
  }

  object El {
    def apply[A](input : => Chunk[A], issues : => List[Issue]) : Input[A] = scalaz.IterV.El[In[A]]((input, issues))
    def apply[A](input : => Chunk[A]) : Input[A] = scalaz.IterV.El[In[A]]((input, Nil))
    def unapply[A](r : Input[A]): Option[In[A]] = scalaz.IterV.El.unapply(r)
  }

  object Done {
    def apply[A, V](valueOption : Option[V], issues : List[Issue], i: => Input[A]) : Iteratee[A,V] = scalaz.IterV.Done((valueOption, issues),i)
    def unapply[A, V](r: Iteratee[A, V]): Option[((Option[V], List[Issue]), Input[A])] = scalaz.IterV.Done.unapply(r)
  }

  object Success {
    def apply[A, V](value : V, issues : List[Issue], i: => Input[A]) : Iteratee[A,V] = {
      val optionValue : Option[V] = Some(value)
      scalaz.IterV.Done((optionValue, issues),i)
    }
    def unapply[A, V](r: Iteratee[A, V]): Option[(V, List[Issue], Input[A])] = {
        r.fold[Option[(V,List[Issue],Input[A])]](
        done = {
          (result, i) => {
            val (valueOption, issues) = result
            if(valueOption.isDefined)
              Some((valueOption.get, issues, i))
            else
              None
          }
        },
        cont = f => None)
    }
  }

  object Failure {
    def apply[A, V](issues : List[Issue], i: => Input[A]) : Iteratee[A,V] = scalaz.IterV.Done((None, issues),i)
    def unapply[A, V](r: Iteratee[A, V]): Option[(List[Issue], Input[A])] = {
      r.fold[Option[(List[Issue],Input[A])]](
        done = {
          (result, i) => {
            val (valueOption, issues) = result
            if(valueOption.isEmpty)
              Some((issues, i))
            else
              None
          }
        },
        cont = f => None)
    }
  }

  object Cont {
    def apply[A, V](f: Input[A] => Iteratee[A,V]) : Iteratee[A,V] = scalaz.IterV.Cont.apply(f)
    def unapply[A, V](r: Iteratee[A,V]): Option[Input[A] => Iteratee[A,V]] = scalaz.IterV.Cont.unapply(r)
  }

  trait Enumerator[A] {
    def enumerate[V](i : Iteratee[A,V]) : Iteratee[A,V]
  }

  trait Enumeratee[A,B] {
    def transform[V](i : Iteratee[B,V]) : Iteratee[A,V]
  }
}

File: api/src/main/java/org/gtri/util/iteratee/api/Iteratee.java
/*
    Copyright 2012 Georgia Tech Research Institute

    Author: lance.gatlin@gtri.gatech.edu

    This file is part of org.gtri.util.iteratee library.

    org.gtri.util.iteratee library is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    org.gtri.util.iteratee library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with org.gtri.util.iteratee library. If not, see <http://www.gnu.org/licenses/>.

*/

package org.gtri.util.iteratee.api;

/**
 * An interface for an Iteratee. An Iteratee is a Mealy machine, a finite state 
 * machine whose output is determined by both its current state and its input. 
 * 
 * @author lance.gatlin@gmail.com
 * @param <I> the input type
 * @param <O> the output type
 */
public interface Iteratee<I,O> {
  /**
   * Get the initial state of the Iteratee
   * @return the initial state of the Iteratee
   */
  Iteratee.State<I,O> initialState();
  
  /**
   * An interface for an immutable state of an Iteratee. Applying a buffer 
   * of input to the Iteratee returns an immutable result object that contains 
   * the next immutable Iteratee state, the output, any overflow (unused input) 
   * and any issues raised during processing. An Iteratee signals completion (or 
   * the presence of errors) by way of its status method. Further calls to the 
   * apply method after an Iteratee's status method has indicated completion are 
   * a noop and will result in all applied input being returned as overflow in 
   * the result. Once all input has been applied to an Iteratee, the endOfInput 
   * method must be called to signal to the Iteratee to move to the SUCCESS or 
   * FATAL_ERROR status.
   * 
   * Note1: Iteratee.State implementations should not throw exceptions. Instead,
   * exceptions should be returned as issues in the result and the 
   * status of the next Iteratee set to FATAL_ERROR should processing be stopped
   * or RECOVERABLE_ERROR if processing may continue.
   * 
   * Note2: Application of input to an Iteratee may result in more overflow than 
   * input applied in the method call. Earlier Iteratee.States may internally 
   * cache input that is then returned by a later Iteratee state.
   * 
   * @author lance.gatlin@gmail.com
   * @param <I> the input type
   * @param <O> the output type
   */
  public static interface State<I,O> {
    /**
     * Get the status of the Iteratee
     * @return status of the Iteratee
     */
    StatusCode statusCode();

    /**
     * Apply input to the Iteratee
     * @param input 
     * @return an immutable result object
     */  
    Iteratee.State.Result<I,O> apply(ImmutableBuffer<I> input);
    
    /**
     * Apply input to the Iteratee
     * @param input 
     * @return an immutable result object
     */  
    Iteratee.State.Result<I,O> apply(I input);
    
    /**
     * Signal to the Iteratee the end of input
     * @return an immutable result object
     */
    Iteratee.State.Result<I,O> endOfInput();
    
    /**
     * The immutable result of applying input to an Iteratee
     * 
     * @param <I> the input type
     * @param <O> the output type
     */  
    public static interface Result<I,O> {
      /**
       * Get the next immutable state of the Iteratee
       * @return the next immutable state of the Iteratee
       */
      Iteratee.State<I,O> next();
      /**
       * Get the output
       * @return the output
       */
      ImmutableBuffer<O> output();
      /**e
       * Get any unused input. May return input cached by previous states.
       * @return any unused input
       */
      ImmutableBuffer<I> overflow();
      /**
       * Get issues raised during processing
       * @return issues raised during processing
       */
      ImmutableBuffer<Issue> issues();
    }
  }
  
}
